src/main/java/com/salesforce/dynamodbv2/mt/util/RecordsCache.java

```diff
@@ -0,0 +1,276 @@
+                return records;
+            }
+
+            return records.subList(getIndex(from), records.size());
+        }
+
+        private int getIndex(SequenceNumber sequenceNumber) {
+            final List<SequenceNumber> sequenceNumbers = Lists.transform(records, SequenceNumber::fromRecord);
+            int index = Collections.binarySearch(sequenceNumbers, sequenceNumber);
+            if (index < 0) {
+                index = (-index) - 1;
+            }
+            return index;
+        }
+
+        Segment subSegment(Segment previousSegment, Segment nextSegment) {
+            return subSegment(
+                previousSegment == null ? null : previousSegment.getEnd().next(),
+                nextSegment == null ? null : nextSegment.getStart()
+            );
+        }
+
+        Segment subSegment(SequenceNumber from, SequenceNumber to) {
+            assert from == null || to == null || from.compareTo(to) < 0;
+
+            int cf = from == null ? 1 : start.compareTo(from);
+            int cl = to == null ? -1 : end.compareTo(to);
+
+            if (cf >= 0) {
+                // first sequence number of this segment is after from: start with first
+                if (cl <= 0) {
+                    // last sequence number of this segment is before to: end with last
+                    return this;
+                } else {
+                    // last sequence number of this segment is after to: end with to
+                    final List<Record> newRecords = new ArrayList<>(records.subList(0, getIndex(to)));
+                    return new Segment(start, to, newRecords);
+                }
+            } else {
+                // first sequence number of this segment if before from: start with from
+                if (cl <= 0) {
+                    // last sequence number of this segment is before to: end with last
+                    final List<Record> newRecords = new ArrayList<>(records.subList(getIndex(from), records.size()));
+                    return new Segment(from, end, newRecords);
+                } else {
+                    // last sequence number of this segment is after to: end with to
+                    final List<Record> newRecords = new ArrayList<>(records.subList(getIndex(from), getIndex(to)));
+                    return new Segment(from, to, newRecords);
+                }
+            }
+        }
+
+        long getByteSize() {
+            return byteSize;
+        }
+
+        boolean isEmpty() {
+            return start.equals(end);
+        }
+    }
+
+    // config parameter
+    private final long maxRecordsByteSize;
+
+    // cached record segments sorted by sequence number within each shard
+    private final Map<ShardId, NavigableMap<SequenceNumber, Segment>> segments;
+    // Insertion order of cache segments for eviction purposes
+    private final LinkedList<ShardLocation> insertionOrder;
+    // looks for mutating lock
+    private final ReadWriteLock lock;
```

Comment:
this is a JVM global lock for any mutations to cached segments? or is it more granular?

Reply:
Oh, I see, given we have a max recordsByteSize across this cache, we have to enforce a JVM global lock, so insertions into the cache properly evict the oldest segments inserted into the cache, across all tenants and shards. 

I'm worried this cache will get so much churn from typical stream processing and delta API requests, that the write lock will begin throttling readers of the cache. Another snippet worth instrumenting to return back to ZOS, as a potential bottleneck of delta API queries.

Reply:
Great questions and observations! Exactly the discussion I was hoping to have with you guys.

In this new version, there are two locks: one is a shard-level lock (https://github.com/salesforce/mt-dynamo/pull/424/files#diff-90d28430b63aa6120b6b0ce714d31748R548) and one is a cache-level lock (this one).

The shard-level lock is not strictly necessary and we currently do not use one in master (see https://github.com/salesforce/mt-dynamo/blob/master/src/main/java/com/salesforce/dynamodbv2/mt/util/CachingAmazonDynamoDbStreams.java#L749). As stated in the comment (https://github.com/salesforce/mt-dynamo/pull/424/files#diff-90d28430b63aa6120b6b0ce714d31748R542), the purpose of the shard-level lock is to sequence access to each underlying shard to avoid hitting DynamoDB limits or throttling (https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html#limits-dynamodb-streams)  and to avoid having concurrent readers (KCL or delta API) load overlapping segments into the cache. There are probably some alternative solutions for addressing each of these concerns (and it may be possible/preferable to address them separately). For example, may it would be sufficient to use a more granular lock at the level of a shard sequence number (or maybe range?) to avoid unnecessary concurrent loads and then just rely on DynamoDB streams (like we do in master) to sequence/throttle access to the shard.

The cache lock guards (1) the non-thread safe collections used by the cache implementation and (2) the invariant that no two cache segments overlap. The invariant is enforced, but checking both the pervious and next cache segments when adding a new one and dropping any reducing the range of the added segment to avoid overlap (https://github.com/salesforce/mt-dynamo/pull/424/files#diff-144e0b563c0b1615fa0417b18acc0778R230). That's what allows the `get` method to simply lookup the next segment by using the end of the current segment (https://github.com/salesforce/mt-dynamo/pull/424/files#diff-144e0b563c0b1615fa0417b18acc0778R195). Again, there are likely different ways to go about implementing this range cache. One of the main motivations for splitting this logic into a separate class is to make it easier for us to change it independently of the caching streams adapter. For example, one option we could explore is to relax the invariant of having no overlap. And then maybe we could use a standard Guava cache with a sorted index we maintain via an eviction lister to reduce the write contention. The assumption underlying the current implementation is that the time to update the cache data structure (i.e., the `putRecords` method) is orders of magnitude faster than loading records from the DynamoDB stream. And since we need to sequence stream reads to some extent anyway to avoid hitting limits, write contention on the in-memory data structure may not be as critical. So I went with a simple solution as a starting point. Definitely worth instrumenting the wait times though to validate that assumption.